modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -235,6 +258,79 @@ private CompletableFuture<Object> processScanRetrieveBatchAction(ReadWriteScanRe
         });
     }
 
+    /**
+     * Process transaction finish request:
+     * <ol>
+     *     <li>Evaluate commit timestamp.</li>
+     *     <li>Run specific raft {@code FinishTxCommand} command, that will apply txn state to corresponding txStateStorage.</li>
+     *     <li>Send cleanup requests to all enlisted primary replicas.</li>
+     * </ol>
+     * This operation is NOT idempotent, because of commit timestamp evaluation.
+     *
+     * @param request Transaction finish request.
+     * @return future result of the operation.
+     */
+    private CompletableFuture<Object> processTxFinishAction(TxFinishReplicaRequest request) {
+        HybridTimestamp commitTimestamp = hybridClock.now();
+
+        List<String> aggregatedGroupIds = request.groups().values().stream().flatMap(List::stream).collect(Collectors.toList());
+
+        UUID txId = request.txId();
+
+        boolean commit = request.commit();
+
+        CompletableFuture<Object> chaneStateFuture = raftClient.run(
+                new FinishTxCommand(
+                        txId,
+                        commit,
+                        commitTimestamp,
+                        aggregatedGroupIds
+                )
+        );
+
+        // TODO: https://issues.apache.org/jira/browse/IGNITE-17578
+        chaneStateFuture.thenRun(
+                () -> request.groups().forEach(
+                        (recipientNode, replicationGroupIds) -> txManager.cleanup(
+                                recipientNode,
+                                replicationGroupIds,
+                                txId,
+                                commit,
+                                commitTimestamp
+                        )
+                )
+        );
+
+        return chaneStateFuture;
+    }
+
+
+    /**
+     * Process transaction cleanup request:
+     * <ol>
+     *     <li>Run specific raft {@code TxCleanupCommand} command, that will convert all pending entries(writeIntents)
+     *     to either regular values(TxState.COMMITED) or removing them (TxState.ABORTED).</li>
+     *     <li>Release all locks that were held on local Replica by given transaction.</li>
+     * </ol>
+     * This operation is idempotent, so it's safe to retry it.
+     *
+     * @param request Transaction cleanup request.
+     * @return CompletableFuture of void.
+     */
+    private CompletableFuture<Object> processTxCleanupAction(TxCleanupReplicaRequest request) {
```

Comment:
May be CompletableFuture<Void> is more sutable.

Reply:
Agreed. Fixed.